File: services/signapps-meet/src/models/remote.rs
//! Remote desktop connection models (absorbed from signapps-remote).

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize, FromRow, utoipa::ToSchema)]
/// Represents a remote connection.
pub struct RemoteConnection {
    pub id: Uuid,
    pub hardware_id: Option<Uuid>,
    pub name: String,
    pub protocol: String, // 'rdp', 'vnc', 'ssh'
    pub hostname: String,
    pub port: i32,
    pub username: Option<String>,
    #[serde(skip_serializing)]
    #[allow(dead_code)]
    pub password_encrypted: Option<String>,
    #[serde(skip_serializing)]
    #[allow(dead_code)]
    pub private_key_encrypted: Option<String>,
    pub parameters: Option<serde_json::Value>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize, Deserialize, utoipa::ToSchema)]
/// Request payload for CreateConnection operation.
pub struct CreateConnectionRequest {
    pub hardware_id: Option<Uuid>,
    pub name: String,
    pub protocol: String,
    pub hostname: String,
    pub port: i32,
    pub username: Option<String>,
    pub password: Option<String>,
    pub private_key: Option<String>,
    pub parameters: Option<serde_json::Value>,
}

#[derive(Debug, Serialize, Deserialize, utoipa::ToSchema)]
/// Request payload for UpdateConnection operation.
pub struct UpdateConnectionRequest {
    pub name: Option<String>,
    pub protocol: Option<String>,
    pub hostname: Option<String>,
    pub port: Option<i32>,
    pub username: Option<String>,
    pub password: Option<String>,
    pub private_key: Option<String>,
    pub parameters: Option<serde_json::Value>,
}

#[derive(Debug, Serialize, Deserialize)]
#[allow(dead_code)]
/// Represents a guacamole instruction.
pub struct GuacamoleInstruction {
    pub opcode: String,
    pub args: Vec<String>,
}

File: client/src/lib/api/remote.ts
/**
 * Remote API Module
 *
 * Migrated to use API Factory pattern.
 * @see factory.ts for client creation details
 */
import { getClient, getServiceBaseUrl, ServiceName } from "./factory";

// Get the remote service client (cached)
const remoteClient = getClient(ServiceName.REMOTE);
const REMOTE_URL = getServiceBaseUrl(ServiceName.REMOTE);

// ============================================================================
// Types
// ============================================================================

export interface RemoteConnection {
  id: string;
  hardware_id?: string;
  name: string;
  protocol: "rdp" | "vnc" | "ssh" | "telnet";
  hostname: string;
  port: number;
  username?: string;
  parameters?: Record<string, unknown>;
  recording_enabled?: boolean;
  recording_path?: string;
  created_at?: string;
  updated_at?: string;
}

export interface CreateConnectionRequest {
  hardware_id?: string;
  name: string;
  protocol: "rdp" | "vnc" | "ssh" | "telnet";
  hostname: string;
  port: number;
  username?: string;
  password?: string;
  private_key?: string;
  parameters?: Record<string, unknown>;
}

export interface UpdateConnectionRequest {
  name?: string;
  protocol?: "rdp" | "vnc" | "ssh" | "telnet";
  hostname?: string;
  port?: number;
  username?: string;
  password?: string;
  private_key?: string;
  parameters?: Record<string, unknown>;
  recording_enabled?: boolean;
}

// ============================================================================
// Remote API
// ============================================================================

export const remoteApi = {
  // List all remote connections
  listConnections: () =>
    remoteClient.get<RemoteConnection[]>("/remote/connections"),

  // Get a single connection
  getConnection: (id: string) =>
    remoteClient.get<RemoteConnection>(`/remote/connections/${id}`),

  // Create a new connection
  createConnection: (data: CreateConnectionRequest) =>
    remoteClient.post<RemoteConnection>("/remote/connections", data),

  // Update a connection
  updateConnection: (id: string, data: UpdateConnectionRequest) =>
    remoteClient.put<RemoteConnection>(`/remote/connections/${id}`, data),

  // Delete a connection
  deleteConnection: (id: string) =>
    remoteClient.delete(`/remote/connections/${id}`),

  // Get WebSocket URL for remote connection (Guacamole protocol)
  getWebSocketUrl: (connectionId: string): string => {
    const wsBaseUrl = REMOTE_URL.replace(/^http/, "ws");
    return `${wsBaseUrl}/remote/ws/${connectionId}`;
  },
};
